### PR TITLE
fix: do not activate extension if config is not found

### PR DESCRIPTION
**Pull request by DudaGod**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,7 +4,7 @@ node_modules
 .DS_Store
 vscode.d.ts
 .eslintcache
-samples/**/.*
+samples/**/.testplane
```
Comment (DudaGod) [inline .gitignore:7]: In order to not ignore `.vscode` folder

File: package.json [hunk 2]
```diff
@@ -38,6 +38,16 @@
           "name": "Testplane"
         }
       ]
+    },
+    "configuration": {
```
Comment (DudaGod) [inline package.json:42]: The options described in this section are displayed in settings interface

File: package.json [hunk 3]
```diff
@@ -50,9 +60,11 @@
     "build": "tsup --minify --clean",
     "watch": "tsup --watch --sourcemap",
     "test": "npm run lint && npm run test-e2e",
-    "test-e2e": "npm run test-e2e:basic && npm run test-e2e:settings-view",
+    "test-e2e": "npm run test-e2e:basic && npm run test-e2e:settings-view && npm run test-e2e:vscode-settings && npm run test-e2e:empty",
     "test-e2e:basic": "wdio run ./tests/e2e/basic/wdio.conf.ts",
     "test-e2e:settings-view": "wdio run ./tests/e2e/settings-view/wdio.conf.ts",
+    "test-e2e:vscode-settings": "wdio run ./tests/e2e/vscode-settings/wdio.conf.ts",
```
Comment (DudaGod) [inline package.json:66]: Here I test ability to use `configPath` specified in `.vscode/settings.json`

File: package.json [hunk 4]
```diff
@@ -50,9 +60,11 @@
...
+    "test-e2e:empty": "wdio run ./tests/e2e/empty/wdio.conf.ts",
```
Comment (DudaGod) [inline package.json:67]: Here I test that extension do nothing on open testing bar

File: src/api/index.ts [hunk 5]
```diff
@@ -3,15 +3,14 @@ import { fork, type ChildProcess } from "node:child_process";
...
 import { WorkerEvents, type WorkerInitFailEventMessage } from "../worker/types";
 import { MasterEvents } from "../types";
+import type { VSCodeConfig } from "../config";
 
 export * from "./rpc";
 
-export async function createChildProcess(wf: vscode.WorkspaceFolder): Promise<ChildProcess> {
+export async function createChildProcess(wf: vscode.WorkspaceFolder, config: VSCodeConfig): Promise<ChildProcess> {
     const workerPath = resolve(__dirname, "worker.js");
-    const configPaths = await vscode.workspace.findFiles(CONFIG_GLOB, CONFIG_GLOB_EXCLUDE);
```
Comment (DudaGod) [inline src/api/index.ts:14]: Now config read only in one place and only if user not specify `configPath` in `.vscode/settings.json`

File: src/config/settings.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import path from "node:path";
+import os from "node:os";
+import * as vscode from "vscode";
+
+export type VSCodeSettings = {
+    configPath: string | undefined;
+};
+
+export const getVSCodeSettings = (wf?: vscode.WorkspaceFolder): VSCodeSettings => {
+    const rootConfig = vscode.workspace.getConfiguration("testplane");
```
Comment (DudaGod) [inline src/config/settings.ts:10]: get global user settings for testplane

File: src/config/settings.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    const folderConfig = vscode.workspace.getConfiguration("testplane", wf);
```
Comment (DudaGod) [inline src/config/settings.ts:11]: get workspace user settings for testplane

File: src/config/settings.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    const get = <T>(key: string, defaultValue?: T): T | undefined =>
```
Comment (DudaGod) [inline src/config/settings.ts:13]: just a helper to more convenient get value from settings

File: src/config/settings.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+type SettingValueOptions<T> = {
+    rootConfig: vscode.WorkspaceConfiguration;
+    folderConfig: vscode.WorkspaceConfiguration;
+    key: string;
+    defaultValue?: T;
+};
+
+function getSettingValue<T>({ rootConfig, folderConfig, key, defaultValue }: SettingValueOptions<T>): T | undefined {
+    return folderConfig.get(key) ?? rootConfig.get(key) ?? defaultValue;
```
Comment (DudaGod) [inline src/config/settings.ts:36]: first of all use settings from workspace, then global and at the default value

File: src/config/utils.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import path from "node:path";
+import * as vscode from "vscode";
+import _ from "lodash";
+import { CONFIG_GLOB, CONFIG_GLOB_EXCLUDE } from "../constants";
+
+export const findTpnConfigFile = async (): Promise<string | undefined> => {
+    const configs = (await vscode.workspace.findFiles(CONFIG_GLOB, CONFIG_GLOB_EXCLUDE)).sort((a, b) => {
+        return a.fsPath.split(path.sep).length - b.fsPath.split(path.sep).length;
```
Comment (DudaGod) [inline src/config/utils.ts:9]: Here I sort found configs in order to use first which is less nested

File: src/extension.ts [hunk 11]
```diff
@@ -87,7 +88,16 @@ class TestplaneExtension {
 
             // TODO: should support few workspaces
             const wf = folders[0];
-            const childProc = await createChildProcess(wf);
+            const config = await getVSCodeConfig(wf);
+
+            if (!config.configPath) {
```
Comment (DudaGod) [inline src/extension.ts:93]: If I don't found config then I just send message to output testplane channel and delete `Reading Testplane tests...` from testing sidebar. After that sidebar looks like as default view.

File: tests/e2e/basic/tsconfig.json [hunk 12]
```diff
@@ -1,17 +1,4 @@
...
-        "noEmit": true,
-        "allowImportingTsExtensions": true,
-        "resolveJsonModule": true,
-        "isolatedModules": true,
-        "strict": true,
-        "noUnusedLocals": true,
-        "noUnusedParameters": true,
-        "noFallthroughCasesInSwitch": true
-    },
+    "extends": "../tsconfig.common.json",
```
Comment (DudaGod) [inline tests/e2e/basic/tsconfig.json:2]: just refactor the same tsconfig files

File: tests/e2e/basic/wdio.conf.ts [hunk 13]
```diff
@@ -1,47 +1,8 @@
...
-    afterTest: async (test, _, { passed }) => {
-        if (passed) {
-            return;
-        }
-
-        const screenshotDir = path.join(__dirname, "..", "screens-on-fail");
-        await fs.mkdir(screenshotDir, { recursive: true });
-        await browser.saveScreenshot(path.join(screenshotDir, `${test.parent} - ${test.title}.png`));
-    },
-};
```
Comment (DudaGod) [inline tests/e2e/basic/wdio.conf.ts:47]: just refactor the same wdio configs

File: tests/e2e/empty/specs/activate.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    it("should do not try to read tests", async () => {
+        const vscodePO = await VSCodePO.create();
+        const testingViewControl = vscodePO.getTestingViewControl();
+        await testingViewControl.open();
+
+        const sidebar = vscodePO.getTestingSideBar();
+        const [firstSection] = await sidebar.getSections();
+        const items = await firstSection.getVisibleItems();
+
+        expect(items).toHaveLength(0);
```
Comment (DudaGod) [inline tests/e2e/empty/specs/activate.ts:14]: Here I check that nothing render inside `Test explorer`. It means that `Reading Testplane tests` is removed

File: tests/e2e/empty/specs/activate.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+        const webview = await vscodePO.getWebviewByTitle("Testplane");
+        expect(webview.elem).toBeDisplayed();
```
Comment (DudaGod) [inline tests/e2e/empty/specs/activate.ts:17]: Check that Testplane webview displayed.

File: tests/e2e/vscode-settings/specs/use-vscode-settings.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        const sidebar = vscodePO.getTestingSideBar();
+        await sidebar.waitTestsRead();
+
+        const [firstSection] = await sidebar.getSections();
+        const [mainTreeItem] = await firstSection.getVisibleItems();
+
+        await mainTreeItem.expandAll();
+        const testsFullTitle = await mainTreeItem.getTestsFullTitle();
+
+        expect(testsFullTitle).toEqual(["tests test.testplane.ts success another-bro"]);
```
Comment (DudaGod) [inline tests/e2e/vscode-settings/specs/use-vscode-settings.ts]: check that used browser name from config that specified in `.vscode/settings.json`

File: package.json (hunk 3 above)
Comment (KuznetsovRoman) [reply package.json:66]: Maybe we should use "concurrently" npm package here in order to run these in parallel?
Example: https://github.com/gemini-testing/testplane/blob/master/package.json#L15

File: src/config/index.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
+import * as vscode from "vscode";
+import { getVSCodeSettings, type VSCodeSettings } from "./settings";
+import { findTpnConfigFile } from "./utils";
```
Comment (KuznetsovRoman) [inline src/config/index.ts]: What is "Tpn"?
Lets use full name here

File: src/config/settings.ts [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+    const get = <T>(key: string, defaultValue?: T): T | undefined =>
+        getSettingValue<T>({
+            rootConfig,
+            folderConfig,
+            key,
+            defaultValue,
+        });
+
+    const configPath = get<string | undefined>("configPath");
```
Comment (KuznetsovRoman) [inline src/config/settings.ts]: Do you need to specify `| undefined` here? `get` already returns `T | undefined`, so it should be `string | undefined` if you only specify `<string>`

File: src/config/utils.ts (hunk 10 above)
Comment (KuznetsovRoman) [reply src/config/utils.ts:9]: Lets write it in more readable format:
```ts
const configs = await vscode.workspace.findFiles(CONFIG_GLOB, CONFIG_GLOB_EXCLUDE);
const configsSorted = configs.sort((a,b) => {
    return a.fsPath.split(path.sep).length - b.fsPath.split(path.sep).length;
})
```

File: src/config/utils.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+    });
+
+    return _.isEmpty(configs) ? undefined : configs[0].fsPath;
```
Comment (KuznetsovRoman) [inline src/config/utils.ts]: > Don't use the undefined primitive. It's a value that the JS compiler will automatically set for you when you declare variables without assignment or if you try to access properties of objects for which there is no reference. On the other hand, use null if and only if you intentionally want a variable to have "no value".

So i would suggest to use `null` here.
Or, at least, get rid of explicit "undefined":
```ts
if (_.isEmpty(configs)) {
    return;
}

return configs[0].fsPath;
```

File: tests/e2e/vscode-settings/specs/use-vscode-settings.ts (hunk 16 above)
Comment (KuznetsovRoman) [reply tests/e2e/vscode-settings/specs/use-vscode-settings.ts]: Weak test, imho
I would just break default testplane config, and here tested that testplane config is successfully read

**Review by KuznetsovRoman — APPROVED** [review]

File: package.json (hunk 3 above)
Comment (DudaGod) [reply package.json:66]: Yeah, I know about this package. But when I run tests in CI in parallel (run few vscode instances at the same time) tests starts to fail (I checked it with few instances in the same wdio). But I didn't check it when run a lot of wdio instances in parallel.

I will try to use it, but most likely there will be a lot of test failures.

File: src/config/index.ts (hunk 17 above)
Comment (DudaGod) [reply src/config/index.ts]: Tpn -> Testplane

**Pull request merged**